storage/redis/redis.go

```diff
@@ -0,0 +1,183 @@
+	"fmt"
+	"strconv"
+
+	"github.com/lovoo/goka/storage"
+
+	redis "gopkg.in/redis.v5"
+)
+
+const (
+	offsetKey = "__offset"
+)
+
+type redisStorage struct {
+	client *redis.Client
+	hash   string
+}
+
+// NewRedis creates a new Storage backed by Redis.
+func NewRedis(client *redis.Client, hash string) (storage.Storage, error) {
+	if client == nil {
+		return nil, errors.New("invalid redis client")
+	}
+	if err := client.Ping().Err(); err != nil {
+		return nil, err
+	}
+	return &redisStorage{
+		client: client,
+		hash:   hash,
+	}, nil
+}
+
+func (s *redisStorage) Has(key string) (bool, error) {
+	return s.client.HExists(s.hash, key).Result()
+}
+
+func (s *redisStorage) Get(key string) ([]byte, error) {
+	has, err := s.client.HExists(s.hash, key).Result()
+	if err != nil {
+		return nil, fmt.Errorf("error checking for existence in redis (key %s): %v", key, err)
+	} else if !has {
+		return nil, nil
+	}
+	value, err := s.client.HGet(s.hash, key).Bytes()
+	if err != nil {
+		return nil, fmt.Errorf("error getting from redis (key %s): %v", key, err)
+	}
+	return value, nil
+}
+
+func (s *redisStorage) GetOffset(defValue int64) (int64, error) {
+	data, err := s.Get(offsetKey)
+	if err != nil {
+		return 0, err
+	}
+	if data == nil {
+		return defValue, nil
+	}
+
+	value, err := strconv.ParseInt(string(data), 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("error decoding redis offset (%s): %v", string(data), err)
+	}
+	return value, nil
+}
+
+func (s *redisStorage) Set(key string, value []byte) error {
+	err := s.client.HSet(s.hash, key, value).Err()
+	if err != nil {
+		return fmt.Errorf("error setting to redis (key %s): %v", key, err)
+	}
+	return nil
+}
+
+func (s *redisStorage) SetOffset(offset int64) error {
+	return s.Set(offsetKey, []byte(strconv.FormatInt(offset, 10)))
+}
+
+func (s *redisStorage) Delete(key string) error {
+	return s.client.HDel(s.hash, key).Err()
+}
+
+func (s *redisStorage) Iterator() (storage.Iterator, error) {
+	var current uint64
+	var keys []string
+	var err error
+
+	keys, current, err = s.client.HScan(s.hash, current, "", 0).Result()
+	if err != nil {
+		return nil, err
+	}
+	return &redisIterator{
+		current: current,
+		keys:    keys,
+		client:  s.client,
+		hash:    s.hash,
+	}, nil
+}
+
+func (s *redisStorage) IteratorWithRange(start, limit []byte) (storage.Iterator, error) {
+	var current uint64
+	var keys []string
+	var err error
+
+	keys, current, err = s.client.HScan(s.hash, current, string(start), 0).Result()
+	if err != nil {
+		return nil, err
+	}
+	return &redisIterator{
+		current: current,
+		keys:    keys,
+		client:  s.client,
+		hash:    s.hash,
+	}, nil
+}
+
+func (s *redisStorage) Recovered() bool {
+	return false
+}
+
+func (s *redisStorage) MarkRecovered() error {
+	return nil
+}
+
+func (s *redisStorage) Open() error {
+	return nil
+}
+
+// Close removes all data present current hash(`namespace:table:partition`)
+// instead closing the redis client.
+func (s *redisStorage) Close() error {
```

Comment:
`Close` should close the client and leave the data in Redis. Deleting the data on close would force the applications to recover their state on every launch and this is not in line with the other storage implementations. Functionality for cleaning the state from Redis would definitely be a convenient addition but I would move it to a separate method.

Reply:
I don't know why it is deleting the data, but I don't think that `Close` should actually close the client too. The caller should care about closing the client since the Redis client is passed to the storage builder and all partitions share the same client.

Reply:
I completely overlooked that part. Even if it wasn't shared with other builders, it would be better to leave it to caller who built the client.

Reply:
I put this `delete` here, 'cause I thought that when you close a partition, you would not use it anymore by this consumer.